### PR TITLE
Add support for git command line, shallow clones, multiple-path monitoring

### LINKED ISSUE #4: Add more test to JGitHelper

### LINKED ISSUE #13: Support for multiple paths?
I've tested out this plugin and it's working great, except my source repo is split up [according to Pants preferred way][pants]. Meaning that my source code lives in `src/<lang>/company/<project>/etc` and tests in `tests/<lang>/company/<project>/etc`. 

Does this plugin support multiple paths to be specified? 
I tried looking through the source, but I couldn't figure it out through GitHub. 
If it doesn't, I wouldn't mind having a stab at adding that feature, but I would love any pointers for how I should approach it. :)

[pants]: https://www.pantsbuild.org/setup_repo.html

### LINKED ISSUE #14: TFS Git support
Unable to connect to TFS Git repo:
`Invalid URL format. Should match ^(?:git|ssh|https?|git@[\w-\.]+[\w]+\.[\w]+):(?:\/\/)?[\w\.@:\/~_-]+\.git(?:\/?|\#[\d\w\.\-_]+?)$`
I think it's because URL format doesn't ends with "*.git"
Is it supported with this plugin?
Also need to authenticate with TFS token or domain user. Is it possible?

### LINKED ISSUE #15: Does not support git ssh
This plugin is only working with https:// style git URLs, and fails to connect with git@ style repository URLs.

### LINKED ISSUE #21: 'Unable to de-serialize...' error when using the git path material plugin
My company has recently moved from svn to git for source control so I have a legacy application that consists of multiple projects which all reside in the same git repository (not ideal I know). Now I have been attempting to use the git path material plugin to setup CI for specific components, however I am getting the bellow error message and due to receiving this error go is repeatedly trying to download the code into the flyweight directory. Anyone else encountering this? Running 20.2.0 on Windows.

(changed the uri path slightly and * out the user id)
[132@MessageListener for MaterialUpdateListener] MaterialDatabaseUpdater:127 - [Material Update] Modification check failed for material: [url=https://gitlab.com/projects/myproject.git, username=*****, path=\server\services\Notification, branch=master]
java.lang.RuntimeException: Unable to de-serialize json response. SCM revision cannot be empty
	at com.thoughtworks.go.plugin.access.scm.JsonMessageHandler1_0.toSCMRevision(JsonMessageHandler1_0.java:312)
	at com.thoughtworks.go.plugin.access.scm.JsonMessageHandler1_0.responseMessageForLatestRevision(JsonMessageHandler1_0.java:156)
	at com.thoughtworks.go.plugin.access.scm.SCMExtension$5.onSuccess(SCMExtension.java:117)
	at com.thoughtworks.go.plugin.access.scm.SCMExtension$5.onSuccess(SCMExtension.java:109)
	at com.thoughtworks.go.plugin.access.PluginRequestHelper.submitRequest(PluginRequestHelper.java:54)
	at com.thoughtworks.go.plugin.access.scm.SCMExtension.getLatestRevision(SCMExtension.java:109)
	at com.thoughtworks.go.server.service.materials.PluggableSCMMaterialPoller.latestModification(PluggableSCMMaterialPoller.java:60)
	at com.thoughtworks.go.server.service.materials.PluggableSCMMaterialPoller.latestModification(PluggableSCMMaterialPoller.java:45)
	at com.thoughtworks.go.server.service.MaterialService.latestModification(MaterialService.java:120)
	at com.thoughtworks.go.server.materials.LegacyMaterialChecker.findLatestModification(LegacyMaterialChecker.java:50)
	at com.thoughtworks.go.server.materials.ScmMaterialUpdater.insertLatestOrNewModifications(ScmMaterialUpdater.java:55)
	at com.thoughtworks.go.server.materials.ScmMaterialUpdater.addNewMaterialWithModifications(ScmMaterialUpdater.java:70)
	at com.thoughtworks.go.server.materials.PluggableSCMMaterialUpdater.addNewMaterialWithModifications(PluggableSCMMaterialUpdater.java:61)
	at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.addNewMaterialWithModifications(MaterialDatabaseUpdater.java:178)
	at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.initializeMaterialWithLatestRevision(MaterialDatabaseUpdater.java:136)
	at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater$1.doInTransaction(MaterialDatabaseUpdater.java:95)
	at com.thoughtworks.go.server.transaction.TransactionCallback.doWithExceptionHandling(TransactionCallback.java:23)
	at com.thoughtworks.go.server.transaction.TransactionTemplate.lambda$executeWithExceptionHandling$2(TransactionTemplate.java:43)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
	at com.thoughtworks.go.server.transaction.TransactionTemplate.executeWithExceptionHandling(TransactionTemplate.java:40)
	at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.updateMaterial(MaterialDatabaseUpdater.java:92)
	at com.thoughtworks.go.server.materials.MaterialUpdateListener.onMessage(MaterialUpdateListener.java:64)
	at com.thoughtworks.go.server.materials.MaterialUpdateListener.onMessage(MaterialUpdateListener.java:32)
	at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.runImpl(JMSMessageListenerAdapter.java:83)
	at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.run(JMSMessageListenerAdapter.java:63)
	at java.base/java.lang.Thread.run(Unknown Source)

Let me take a step back on this and ask if anyone has an example of how to properly configure a git path material?

Using 20.2.0

### LINKED ISSUE #22: Add support for shallow clones
GoCD server allows use of shallow clones on its Git materials on agents which can massively improve clone performance on larger repositories.

### LINKED ISSUE #23: Add support for `--no-checkout` on server-side clones
GoCD Server's main git material does `--no-checkout` clones server side when clone repositories in order to watch them for updates on materials. This support would be useful to make the Git Path a drop-in replacement for the off-the-shelf Git material; since this is a more performant approach server-side when building changelogs and checking for versions.

### PR DESCRIPTION
**Pull request by chadlwilson**
This PR overhauls the plugin adding support for a number of major features that improve performance and flexibility and seek to make it closer to a drop-in replacement for the off-the-shelf Git material packaged with GoCD.

I apologise for the size of this PR; however I couldn't find a way to get this plugin back into shape for 2019 and modern GoCD without major changes. I have rebased commits & grouped in a way that commits should be able to be reviewed independently - but they do build on top of one anotehr.

I have been maintaining a fork for the last 6 months; however I would prefer if we can merge this and then perhaps migrate this plugin to `gocd-contrib` for more longer-term maintenance; as well as update the plugin directory on gocd.org - as this is a really useful plugin.

**FEATURES**
* effectively deprecates use of `JGit`; prefering the `git` command line if it is available on agents. This is effectively how GoCD ships out-of-the-box in 2019+ (and maybe earlier). (fixes #15 ) 
* switches to [git-cmd](https://github.com/ashwanthkumar/git-cmd) for all `Jgit` vs `git-cmd` handling; using an updated version of the shared library supported by @ashwanthkumar (and already used by [gocd-build-github-pull-requests](https://github.com/ashwanthkumar/gocd-build-github-pull-requests)) . This library now internally has support (that I added via PR in `2.0`) for
  * all git-path filtered operations
  * shallow clone handling when used with `git-cmd` using same logic as GoCD server (fixes #22)
  * ability to do `--no-checkout` clones (fixes #23 )
  * removes all the duplicated code and tests between these two libraries which fixes some niggles when using the plugin (fixes #4, fixes #21 )
  * see https://github.com/ashwanthkumar/git-cmd/pull/4
* adds support for multiple paths, comma-delimited (fixes #13 )
* implements plugin-side support for **shallow clones** (fixes #22)
* processes server-side requests using `--no-checkout` clones (fixes #23 )
* removes the requirement for repo URLs to be suffixed with `.git` (fixes #14 )
   * also allows plugin to work with AWS CodeCommit

**QUALITY OF LIFE**
* Adds clarifying notes to the UI for configuring the material (see the `README` screenshots)
* Uses a trick to get the VSM to show up with the regular Git materials icon ;-) (see the `README` screenshots)
* Documents use with the GoCD YAML plugin

**CHORE**
* Adds `LICENSE`. Much of the original code here appears to have been lift+shifted from GoCD server way back in the past; so to me this plugin **must** be distributed under Apache 2
* Makes code a little more idiomatic/simpler on Java9+
* Gradle and dependencies bumped to latest and rationalized
* Switch to the Gradle GoCD plugin helper for releases; similar to most official plugins

**BREAKING**
* Plugin ID changed to be more idiomatic; so this `2.1.0+` version will show up as a separate plugin in GoCD and requires migration to use it.
* Only builds/targets Java9+ from now on.

### Testing
* This has been battle tested in production for over 6 months on GoCD installations between `19.7.0` and `20.1.0` with real development teams using
     * both HTTPS and git ssh clones
     * hundreds of pipelines using dozens of different *Git path* materials
     * both elastic and static agents under Docker/Kubernetes
* Works with the included `docker-compose.yml` and sample pipelines expressed as `yaml`

**Not tested**
* Submodules. This functionality didn't look like it was ever properly implemented; and I have no use-case for it.

### CODE REVIEW

**Comment by jskswamy** [comment]
Thanks for the PR, will check and add my comments if any

**Comment by jskswamy** [comment]
git-path-material

I tried running 'gradle clean startGoCd' and noticed the above error. Gocd is not able to load the plugin

**Comment by chadlwilson** [comment]
> I tried running 'gradle clean startGoCd' and noticed the above error. Gocd is not able to load the plugin

Can you pull and try again? Apologies; I made a late update to the `vendorName` to add  "& contributors" and forgot that the gocd gradle task helpers don't properly escape `&` in the fields when generating the plugin XML.

**Comment by jskswamy** [comment]
@chadlwilson I've fixed an issue with the sample gocd configuration, kindly accept these changes and update this PR again

https://github.com/chadlwilson/gocd-git-path-material-plugin/pull/1/commits

**Comment by jskswamy** [comment]
@chadlwilson lets make necessary changes to https://github.com/TWChennai/gocd-git-path-sample to show the fan-in and fan-out capability of this plugin.

File: README.md [hunk 1]
```diff
@@ -37,16 +123,29 @@ Execute the following gradle task to start the go-server
...
+Released via the [gocd-plugin-gradle-task-helpers](https://github.com/gocd/gocd-plugin-gradle-task-helpers) onto
+GitHub.
+
+* Check the`gocdPlugin.pluginVersion` version in `build.gradle` and bump if necessary
+* Tag and publish
+    ```bash
+    ./gradlew githubRelease                       
+    PRERELEASE=no ./gradlew githubRelease
+    ```
+* Edit the release description on `GitHub` if necessary.
```
Comment (jskswamy) [inline README.md:151]: Can we add a section for Upgrade guide?

The plugin ID got changed from 'GitPathMaterial' to 'git-path', it's worth mentioning it for existing user to use this plugin upgrade to the newer version without much tussle

Comment (chadlwilson) [reply README.md:151]: Good idea; will do that tonight.

Comment (chadlwilson) [reply README.md:167]: I have added this now.

**Comment by chadlwilson** [comment]
> @chadlwilson I've fixed an issue with the sample gocd configuration, kindly accept these changes and update this PR again
> 
> https://github.com/chadlwilson/gocd-git-path-material-plugin/pull/1/commits

I deliberately moved the sample YAML pipeline to this repository, because it took me quite a while hunting around to find that this was actually supported with the GoCD YAML config plugin; and I felt it was important to have a fully-fledged sample of what this config looks like alongside the plugin code.

It also does not seem necessary to hard-configure SCMs in the GoCD server config as the previous one did; these are automatically created by GoCD. Is there a reason I am missing that you feel we need to add them back?

> @chadlwilson lets make necessary changes to https://github.com/TWChennai/gocd-git-path-sample to show the fan-in and fan-out capability of this plugin.

I can submit a PR that adds the `integration` data that is on my fork here: https://github.com/chadlwilson/gocd-git-path-sample - if we leave the pipeline config sample in this repo would that suffice?

**Comment by jskswamy** [comment]
@chadlwilson the only reason I've added them back is, as soon as i run `gradle startGoCd` I should see the gocd configured correctly with the pipeline and necessary agents.

This was when we fix the issue with the plugin we can easily test whether it works without configuring the pipeline manually again and again

**Comment by chadlwilson** [comment]
> @chadlwilson the only reason I've added them back is, as soon as i run `gradle startGoCd` I should see the gocd configured correctly with the pipeline and necessary agents.
> 
> This was when we fix the issue with the plugin we can easily test whether it works without configuring the pipeline manually again and again

Agreed, and it does work with the previous configuration when pointed to my forks for the `sample` repo as I was using it all the time.

I don't think you have described the problem you are having so I'm not sure I understand what didn't work for you. When you ran `gradlew startGocd` with the sample included with the repo, what happened? I can imagine there is an issue with the missing `integration` path I used but that'd be after the pipelines import, gocd starts; and should be easily fixed without changing the pipeline definition.

**Comment by jskswamy** [comment]
@chadlwilson without the changes to `/extras/config.xml` on starting the go cd with `gradle startGoCd` I end up seeing the following screen.

no-pipeline

with the configuration changes I directly see the pipeline and the agent automatically picks up the pipeline and runs it

auto-pipeline

**Comment by chadlwilson** [comment]
Thanks @jskswamy - you're right that it was a bit confusing to review this because of the tangle between this repo and the separate sample repository. I had mostly set up the config so it would all "just work" once everything was merged, but hard to review in advance.

Anyway, now I have
* moved back to having the pipeline YAML defined in the `gocd-git-path-samples` repo as it was before
* submitted [a separate PR](https://github.com/TWChennai/gocd-git-path-sample/pull/1) to modernize this YAML
* validated that this is working fine for me locally on this branch; pointing to my sample branch [1]

If you want to validate that both of these PRs "work together" prior to merging either of them, you will need to clone this repo/branch and then edit `extras/config.xml` and change the URL to
`<git url="https://github.com/chadlwilson/gocd-git-path-sample" branch="prepare-upstream-merge" />` rather than the intended value once this is merged.

This is all you should need to change. When you start GoCD; you will still initially still see the "add a new pipeline" wizard; that's because GoCD displays that by default on a fresh GoCD now, and the config repos and SCMs are loaded/created asynchronously. If you ignore the wizard and click "admin > Config repositories" or "dashboards" after a few seconds you should see it parsing the config repo and loading the pipelines; then firing them off once the agent registers.

image

[1] Note that until the PR is merged you will see an error from the plugin on the **integration** pipeline like this

image

This is just the way the plugin behaves (and behaved in `1.2.3` i believe) when you try and monitor a path that doesn't exist in the repository. [The PR](https://github.com/TWChennai/gocd-git-path-sample/pull/1) to `gocd-git-path-sample` would need to be merged; or the sample pipeline repo/branch altered to avoid this; as it adds a 3rd path to be monitored.

**Comment by jskswamy** [comment]
Looks good to me will merge both the PR.

Thanks a lot for your contribution

**Comment by chadlwilson** [comment]
Awesome, thanks! Are you cool if I cut a release on the main project?

**Comment by sundbry** [comment]
Great work @chadlwilson thanks for all of this 👍 This is indeed a very useful plugin.

**Comment by jskswamy** [comment]
@chadlwilson made you as collaborator for this project, kindly release it on the main project.

**Comment by chadlwilson** [comment]
Thanks @jskswamy - I've made a release [here](https://github.com/TWChennai/gocd-git-path-material-plugin/releases/tag/v2.1.0-78) now; and switched to using it on our project.

Would you mind making me a collaborator on the samples repo too? I'd quite like to add a sample that gives an idea of how to use it for environment side-channels for deployments; and change the samples to allow easy verification that SSH-based usage is working also  :-)

**Comment by jskswamy** [comment]
@chadlwilson made you as collaborator for [gocd-git-path-sample](https://github.com/TWChennai/gocd-git-path-sample) as well

**Pull request merged**